lib/service/service.go

```diff
@@ -3203,8 +3203,25 @@ func (process *TeleportProcess) initDiagnosticService() error {
 	logger.InfoContext(process.ExitContext(), "Starting diagnostic service.", "listen_address", process.Config.DiagnosticAddr.Addr)
 
 	process.RegisterFunc("diagnostic.service", func() error {
-		err := server.Serve(listener)
-		if err != nil && err != http.ErrServerClosed {
+		muxListener, err := multiplexer.New(multiplexer.Config{
+			Context:                        process.GracefulExitContext(),
```

Comment:
```suggestion
			Context:                        process.ExitContext(),
```

(if not outright `context.Background()`? The lifetime of the diag server should really be "until `diagnostic.shutdown` runs", but it's kinda the same as `process.ExitContext()` I think)